### PR TITLE
Introduce new join points

### PR DESCRIPTION
**Pull request by la4ezar**
**Description**
Introduce two new JoinPointLocations - PRE/POST SendNotification and PRE/POST NotificationStatusReturned.

Changes proposed in this pull request:
- Introduce two new JoinPointLocations
- Unit tests

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [ ] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

**Review by nyordanoff — APPROVED** [review]

**Review by dragobt — APPROVED** [review]

File: components/director/internal/domain/formation/notifications_test.go [hunk 1]
```diff
@@ -539,137 +541,166 @@ func Test_NotificationsService_SendNotification(t *testing.T) {
...
+		FormationID: FormationID,
+		ApplicationTemplate: &webhook.ApplicationTemplateWithLabels{
+			ApplicationTemplate: fixApplicationTemplateModel(),
+			Labels:              fixApplicationTemplateLabelsMap(),
+		},
+		Application: &webhook.ApplicationWithLabels{
+			Application: fixApplicationModel(ApplicationID),
+			Labels:      fixApplicationLabelsMap(),
+		},
+		Runtime: &webhook.RuntimeWithLabels{
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formation/notifications_test.go:560]: There is already a fixture for the runtime with labels.

File: components/director/internal/domain/formation/notifications_test.go [hunk 2]
```diff
@@ -539,137 +541,166 @@ func Test_NotificationsService_SendNotification(t *testing.T) {
...
+			Runtime: fixRuntimeModel(RuntimeID),
+			Labels:  fixRuntimeLabelsMap(),
+		},
+		RuntimeContext: nil,
+		CustomerTenantContext: &webhook.CustomerTenantContext{
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formation/notifications_test.go:565]: There is a fixture here as well - `fixCustomerTenantContext`.

File: components/director/internal/domain/formation/service_test.go [hunk 3]
```diff
@@ -2642,6 +2641,13 @@ func TestServiceResynchronizeFormationNotifications(t *testing.T) {
 	nilSchemaLblDef := fixScenariosLabelDefinition(TntInternalID, testSchema)
 	nilSchemaLblDef.Schema = nil
 
+	formationNotificationSyncDeleteExtRequest := &webhookclient.FormationNotificationRequestExt{
+		Request:       formationNotificationSyncDeleteRequest.Request,
+		Operation:     model.DeleteFormation,
+		Formation:     fixFormationModelWithStateAndAssignmentError(t, model.DeleteErrorFormationState, testErr.Error(), 2),
+		FormationType: testFormationTemplateName,
+	}
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formation/service_test.go:2649]: maybe we could extract this in the fixtures as well?

File: components/director/internal/domain/formationassignment/service_test.go [hunk 4]
```diff
@@ -42,6 +47,35 @@ var (
...
+	formationTemplate = &model.FormationTemplate{
+		ID:                     "testFTID",
+		Name:                   "testFTName",
+		ApplicationTypes:       nil,
+		RuntimeTypes:           nil,
+		RuntimeTypeDisplayName: nil,
+		RuntimeArtifactKind:    nil,
+		TenantID:               nil,
+		Webhooks:               nil,
+		LeadingProductIDs:      nil,
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formationassignment/service_test.go:68]: Do you need to specify these as nil? Aren't they by default nil?

File: components/director/internal/domain/formationassignment/service_test.go [hunk 5]
```diff
@@ -33,6 +35,9 @@ var (
 	testErr = errors.New("Test Error")
 
 	faInput = fixFormationAssignmentModelInput(TestConfigValueRawJSON)
+	fa      = fixFormationAssignmentModelWithFormationID("test")
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formationassignment/service_test.go:38]: you could use `TestFormationID` here. Also for the rest of the changes we could check and use already introduced test variableas and if we don't have them - we can create them

File: components/director/internal/domain/formationassignment/service_test.go [hunk 6]
```diff
@@ -42,6 +47,35 @@ var (
...
+	}
+	reverseFa = fixReverseFormationAssignment(fa)
+
+	rtmTypeLabelKey = "rtmTypeLabelKey"
+	appTypeLabelKey = "appTypeLabelKey"
+
+	appLbl = &model.Label{Value: appSubtype}
+
+	source = "source"
+	target = "target"
```
Comment (PetarTodorovv) [inline components/director/internal/domain/formationassignment/service_test.go:78]: `TestSource`, `TestTarget`

File: components/director/pkg/formationconstraint/join_point_details.go [hunk 7]
```diff
@@ -126,3 +126,45 @@ func (d *GenerateFormationNotificationOperationDetails) GetMatchingDetails() Mat
...
+}
+
+// NotificationStatusReturnedOperationDetails contains details applicable to notification status returned join point
+type NotificationStatusReturnedOperationDetails struct {
+	ResourceType               model.ResourceType
+	ResourceSubtype            string
+	Location                   JoinPointLocation
+	Operation                  model.FormationOperation
+	FormationAssignment        *model.FormationAssignment
+	ReverseFormationAssignment *model.FormationAssignment
```
Comment (PetarTodorovv) [inline components/director/pkg/formationconstraint/join_point_details.go:159]: Check the comment above for the webhook.assignment/reverse assignment type

File: components/director/internal/domain/formation/notifications.go [hunk 8]
```diff
@@ -120,15 +127,53 @@ func (ns *notificationsService) GenerateFormationNotifications(ctx context.Conte
...
+	webhookModel, err := ns.webhookConverter.ToModel(&webhookGql)
+	if err != nil {
+		return nil, errors.Wrap(err, "while converting webhook to model")
+	}
+
+	joinPointDetails := &formationconstraint.SendNotificationOperationDetails{
+		ResourceType:               webhookNotificationReq.GetObjectType(),
+		ResourceSubtype:            webhookNotificationReq.GetObjectSubtype(),
+		Operation:                  webhookNotificationReq.GetOperation(),
+		Webhook:                    webhookModel,
```
[the remark this answers is not in the document]
Comment (la4ezar) [reply components/director/internal/domain/formation/notifications.go:166]: I'm not sure what could be the potential usage too, but it was in the task requirements that's why I added it.

File: components/director/internal/domain/formationassignment/service.go [hunk 9]
```diff
@@ -906,6 +960,131 @@ func (s *service) matchFormationAssignmentsWithRequests(ctx context.Context, ass
...
+		if !apperrors.IsNotFoundError(err) {
+			log.C(ctx).Errorf("An error occurred while getting reverse formation assignment: %v", err)
+			return nil, errors.Wrap(err, "An error occurred while getting reverse formation assignment")
+		}
+		log.C(ctx).Debugf("Reverse assignment with source %q and target %q in formation with ID %q is not found.", fa.Target, fa.Source, formation.ID)
+	}
+
+	return &formationconstraint.NotificationStatusReturnedOperationDetails{
+		ResourceType:               model.FormationResourceType,
+		ResourceSubtype:            template.Name,
```
[the remark this answers is not in the document]
Comment (la4ezar) [reply components/director/internal/domain/formationassignment/service.go:987]: I've extracted both PrepareDetailsForNotificationStatusReturned, CreateExtendedFARequest, and getObjectSubtype(as an unexported method) in the formation assignment `notifications` interfaces.

**Pull request merged**